azurerm/internal/services/policyinsights/parse/remediation.go

```diff
@@ -0,0 +1,130 @@
+package parse
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+)
+
+type ProvisioningType int
+
+const (
+	AtSubscription ProvisioningType = iota
+	AtManagementGroup
+	AtResourceGroup
+	AtResource
+)
+
+type RemediationId struct {
+	Name string
+	RemediationScopeId
+}
+
+// TODO: This paring function is currently suppressing every case difference due to github issue: https://github.com/Azure/azure-rest-api-specs/issues/8353.
+// Currently the returned Remediation response from the service will have all the IDs converted into lower cases.
+func RemediationID(input string) (*RemediationId, error) {
+	// in general, the id of a remediation should be:
+	// {scope}/providers/Microsoft.PolicyInsights/remediations/{name}
+	regex := regexp.MustCompile(`/providers/[Mm]icrosoft\.[Pp]olicy[Ii]nsights/remediations/`)
+	if !regex.MatchString(input) {
+		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q", input)
+	}
+
+	segments := regex.Split(input, -1)
+
+	if len(segments) != 2 {
+		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q: Expected 2 segments after splition", input)
+	}
+
+	scope := segments[0]
+	scopeId, err := RemediationScopeID(scope)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q: %+v", input, err)
+	}
+
+	name := segments[1]
+	if name == "" {
+		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q: remediation name is empty", input)
+	}
+
+	id := RemediationId{
+		Name:               name,
+		RemediationScopeId: *scopeId,
+	}
+
+	return &id, nil
+}
+
+type RemediationScopeId struct {
```

Comment:
I'm not sure we should share the properties struct like this. not all properties will be populated for each of the remediation types. While they share a similar base we should have 4 derived types for each type of ID.eg PolicyRemediationAtSubscriptionID` and such

Reply:
I have done some tweaks on the structure of this part. Introduced an interface for `RemediationScopeId` and 4 structs implemented this interface, please have a look again